File: profiles/java-spring/scaffold/src/test/java/com/example/app/TelemetryTest.java
package com.example.app;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;

import com.example.app.Telemetry.MetricKey;
import com.example.app.Telemetry.SpanIds;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.junit.jupiter.api.io.TempDir;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

/**
 * Unit tests for {@link Telemetry} — the span builder, bounded-cardinality Prometheus metrics, and
 * structured JSON logs. Mirrors the go reference's {@code telemetry_test.go} test-for-test so the
 * behavioural contract (schema, status boundary, escaping, cardinality bound) matches across
 * stacks.
 */
class TelemetryTest {

  private static final ObjectMapper JSON = new ObjectMapper();

  @AfterEach
  void resetState() {
    // The counter state is process-global (mirrors go's package-level map); never leak between
    // tests.
    Telemetry.resetMetrics();
  }

  /** Runs {@code action}, redirecting stdout to a buffer, and returns what it printed. */
  private static String captureStdout(Executable action) {
    PrintStream original = System.out;
    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    System.setOut(new PrintStream(buffer, true, StandardCharsets.UTF_8));
    try {
      action.execute();
    } catch (Throwable t) {
      throw new AssertionError("action under capture threw", t);
    } finally {
      System.setOut(original);
    }
    return buffer.toString(StandardCharsets.UTF_8);
  }

  @Test
  void newSpanIdsHasExpectedLengthsAndIsHex() {
    SpanIds ids = Telemetry.newSpanIds();

    assertThat(ids.traceId()).hasSize(32).matches("[0-9a-f]{32}");
    assertThat(ids.spanId()).hasSize(16).matches("[0-9a-f]{16}");
  }

  @Test
  void newSpanIdsIsRandomAcrossDraws() {
    SpanIds first = Telemetry.newSpanIds();
    SpanIds second = Telemetry.newSpanIds();

    assertThat(second.traceId()).isNotEqualTo(first.traceId());
    assertThat(second.spanId()).isNotEqualTo(first.spanId());
  }

  @Test
  void buildSpanSchema() {
    Map<String, String> attrs = Map.of("http.method", "GET", "http.route", "/healthz");
    Map<String, Object> span = Telemetry.buildSpan("GET /healthz", 1000L, 2500L, attrs, 200);

    assertThat(span.get("start_unix_nano")).isEqualTo("1000");
    assertThat(span.get("end_unix_nano")).isEqualTo("2500");
    assertThat(span.get("parent_span_id")).isNull();
    assertThat(span.get("name")).isEqualTo("GET /healthz");
    @SuppressWarnings("unchecked")
    Map<String, String> gotAttrs = (Map<String, String>) span.get("attributes");
    assertThat(gotAttrs).containsEntry("http.route", "/healthz");

    assertThat((String) span.get("trace_id")).hasSize(32);
    assertThat((String) span.get("span_id")).hasSize(16);
  }

  @ParameterizedTest
  @CsvSource({"200,OK", "404,OK", "499,OK", "500,ERROR", "503,ERROR"})
  void buildSpanStatusBoundary(int statusCode, String wantCode) {
    Map<String, Object> span = Telemetry.buildSpan("op", 0L, 1L, Map.of(), statusCode);

    @SuppressWarnings("unchecked")
    Map<String, Object> status = (Map<String, Object>) span.get("status");
    assertThat(status.get("code")).isEqualTo(wantCode);
  }

  @Test
  void emitSpanWritesToFile(@TempDir Path dir) throws Exception {
    Path sink = dir.resolve("spans.ndjson");

    Telemetry.emitSpan(Telemetry.buildSpan("first", 0L, 1L, Map.of(), 200), sink.toString());
    Telemetry.emitSpan(Telemetry.buildSpan("second", 2L, 3L, Map.of(), 500), sink.toString());

    List<String> lines = Files.readAllLines(sink, StandardCharsets.UTF_8);
    assertThat(lines).hasSize(2); // append semantics, not overwrite
    for (String line : lines) {
      assertThatCode(() -> JSON.readTree(line)).doesNotThrowAnyException();
    }
  }

  @Test
  void emitSpanWritesToStdoutWhenSinkUnset() {
    String out =
        captureStdout(
            () ->
                Telemetry.emitSpan(
                    Telemetry.buildSpan("stdout-span", 10L, 20L, Map.of(), 200), ""));

    JsonNode span = parseJsonLine(out);
    assertThat(span.get("name").asText()).isEqualTo("stdout-span");
  }

  @Test
  void emitSpanBadSinkDoesNotThrow(@TempDir Path dir) {
    // A sink under a nonexistent directory cannot be opened; emitSpan must degrade silently
    // rather than throw and break the request path it observes.
    String badSink = dir.resolve("no-such-dir").resolve("spans.ndjson").toString();

    assertThatCode(
            () ->
                Telemetry.emitSpan(Telemetry.buildSpan("dropped", 0L, 1L, Map.of(), 200), badSink))
        .doesNotThrowAnyException();
  }

  @Test
  void emitSpanPublicOverloadDoesNotThrow() {
    // Exercises the env-driven public entry point (the seam the server actually calls).
    assertThatCode(
            () ->
                captureStdout(
                    () ->
                        Telemetry.emitSpan(Telemetry.buildSpan("via-env", 0L, 1L, Map.of(), 200))))
        .doesNotThrowAnyException();
  }

  @Test
  void recordMetricBoundedCardinality() {
    Telemetry.recordMetric("GET", 200, 5);
    Telemetry.recordMetric("BREW", 418, 3); // unknown method -> "other"
    Telemetry.recordMetric("get", 200, 1); // case-sensitive: lowercase is not known -> "other"

    String out = Telemetry.renderMetrics();

    assertThat(out).contains("http_requests_total{method=\"GET\",status=\"200\"} 1");
    assertThat(out).contains("http_requests_total{method=\"other\",status=\"418\"} 1");
    assertThat(out).contains("http_requests_total{method=\"other\",status=\"200\"} 1");
    assertThat(out).doesNotContain("method=\"BREW\"").doesNotContain("method=\"get\"");
  }

  @Test
  void recordMetricCounterAggregates() {
    Telemetry.recordMetric("POST", 201, 2);
    Telemetry.recordMetric("POST", 201, 2);

    String out = Telemetry.renderMetrics();

    assertThat(out).contains("http_requests_total{method=\"POST\",status=\"201\"} 2");
  }

  @Test
  void renderMetricsShape() {
    Telemetry.recordMetric("GET", 200, 5);

    String out = Telemetry.renderMetrics();

    assertThat(out)
        .contains("# HELP http_requests_total Total number of HTTP requests handled.")
        .contains("# TYPE http_requests_total counter")
        .contains(
            "# HELP http_request_duration_seconds_total Total accumulated request duration in"
                + " seconds.")
        .contains("# TYPE http_request_duration_seconds_total counter")
        .contains("http_request_duration_seconds_total ")
        .endsWith("\n");
  }

  @Test
  void renderMetricsDurationAccumulates() {
    Telemetry.recordMetric("GET", 200, 500); // 0.5s
    Telemetry.recordMetric("GET", 200, 250); // 0.25s -> total 0.75s

    String out = Telemetry.renderMetrics();

    assertThat(out).contains("http_request_duration_seconds_total 0.75");
  }

  @Test
  void resetMetricsClearsState() {
    Telemetry.recordMetric("GET", 200, 5);

    Telemetry.resetMetrics();
    String out = Telemetry.renderMetrics();

    assertThat(out).doesNotContain("http_requests_total{");
    assertThat(out).contains("http_request_duration_seconds_total 0");
  }

  @ParameterizedTest
  @CsvSource(
      delimiterString = "|",
      value = {
        "GET|GET",
        "a\"b|a\\\"b",
        "a\\b|a\\\\b",
      })
  void escapeLabelValue(String input, String want) {
    assertThat(Telemetry.escapeLabelValue(input)).isEqualTo(want);
  }

  @Test
  void escapeLabelValueEscapesNewline() {
    assertThat(Telemetry.escapeLabelValue("a\nb")).isEqualTo("a\\nb");
  }

  @Test
  void renderMetricsEscapesLabels() {
    // Defensive: even if a value with quotes/backslashes/newlines reaches the renderer (bypassing
    // the normal recordMetric bounding), it must not break or inject a series line.
    synchronized (Telemetry.METRICS_LOCK) {
      Telemetry.requestsTotal.put(new MetricKey("x\"\\\n", 200), 1L);
    }

    String out = Telemetry.renderMetrics();

    assertThat(out).contains("method=\"x\\\"\\\\\\n\"");
  }

  @Test
  void logShapeHasDefaultServiceName() {
    String out =
        captureStdout(() -> Telemetry.log(Map.of("event", "request", "status", 200), null));

    JsonNode record = parseJsonLine(out);
    assertThat(record.get("level").asText()).isEqualTo("info");
    assertThat(record.get("service").asText()).isEqualTo("reference-app");
    assertThat(record.get("event").asText()).isEqualTo("request");
    assertThatCode(() -> Instant.parse(record.get("ts").asText())).doesNotThrowAnyException();
  }

  @Test
  void logRejectsBlankServiceOverrideFallsBackToDefault() {
    String out = captureStdout(() -> Telemetry.log(Map.of("event", "boot"), ""));

    JsonNode record = parseJsonLine(out);
    assertThat(record.get("service").asText()).isEqualTo("reference-app");
  }

  @Test
  void logServiceNameOverride() {
    String out = captureStdout(() -> Telemetry.log(Map.of("event", "boot"), "custom-svc"));

    JsonNode record = parseJsonLine(out);
    assertThat(record.get("service").asText()).isEqualTo("custom-svc");
  }

  @Test
  void logPublicOverloadDoesNotThrow() {
    // Exercises the env-driven public entry point (the seam the server actually calls).
    assertThatCode(() -> captureStdout(() -> Telemetry.log(Map.of("event", "via-env"))))
        .doesNotThrowAnyException();
  }

  private static JsonNode parseJsonLine(String out) {
    try {
      return JSON.readTree(out.strip());
    } catch (Exception e) {
      throw new UncheckedIOException(
          new IOException("captured output is not valid JSON: " + out, e));
    }
  }
}

File: profiles/kotlin/scaffold/src/test/kotlin/com/example/app/TelemetryTest.kt
package com.example.app

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.nio.file.Files
import java.nio.file.Path
import java.time.Instant
import java.time.format.DateTimeParseException
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for the pure telemetry primitives (spans, bounded-cardinality
 * Prometheus metrics, structured logs). Mirrors the go/python reference test
 * suites (telemetry_test.go / test_telemetry.py) so the shapes stay identical
 * across profiles. [Telemetry.emitSpan]'s `sinkPath` and [Telemetry.log]'s
 * `serviceName` are driven directly (the injectable-default seam) instead of
 * mutating the real process environment, which the JVM does not support.
 */
class TelemetryTest {
    private val mapper: ObjectMapper = jacksonObjectMapper()

    @AfterEach
    fun resetState() {
        Telemetry.resetMetrics()
    }

    /** Redirect stdout for the duration of [fn] and return what it printed. */
    private fun captureStdout(fn: () -> Unit): String {
        val original = System.out
        val buffer = ByteArrayOutputStream()
        System.setOut(PrintStream(buffer, true, "UTF-8"))
        try {
            fn()
        } finally {
            System.setOut(original)
        }
        return buffer.toString("UTF-8")
    }

    @Test
    fun `newSpanIds are hex of the right width`() {
        val (traceId, spanId) = Telemetry.newSpanIds()
        assertEquals(32, traceId.length) // 16 bytes hex
        assertEquals(16, spanId.length) // 8 bytes hex
        traceId.toBigInteger(16) // valid hex (throws otherwise)
        spanId.toBigInteger(16)
    }

    @Test
    fun `newSpanIds are random`() {
        assertNotEquals(Telemetry.newSpanIds(), Telemetry.newSpanIds())
    }

    @Test
    fun `buildSpan has all required keys`() {
        val span =
            Telemetry.buildSpan(
                name = "GET /healthz",
                startUnixNano = 1,
                endUnixNano = 2,
                attributes = mapOf("http.request.method" to "GET"),
                statusCode = 200,
            )
        for (key in listOf(
            "trace_id", "span_id", "parent_span_id", "name",
            "start_unix_nano", "end_unix_nano", "attributes", "status",
        )) {
            assertTrue(span.containsKey(key), "missing key $key")
        }
        assertEquals("GET /healthz", span["name"])
        assertNull(span["parent_span_id"])
        assertEquals(mapOf("http.request.method" to "GET"), span["attributes"])
        val traceId = span["trace_id"] as String
        val spanId = span["span_id"] as String
        assertEquals(32, traceId.length)
        assertEquals(16, spanId.length)
    }

    @Test
    fun `buildSpan nanos are decimal strings`() {
        val span =
            Telemetry.buildSpan(
                name = "x",
                startUnixNano = 1_700_000_000_000_000_000L,
                endUnixNano = 1_700_000_000_500_000_000L,
                attributes = emptyMap(),
                statusCode = 200,
            )
        assertEquals("1700000000000000000", span["start_unix_nano"])
        assertEquals("1700000000500000000", span["end_unix_nano"])
    }

    @Test
    fun `buildSpan status boundary is OK below 500 and ERROR at or above`() {
        val cases = listOf(200 to "OK", 404 to "OK", 499 to "OK", 500 to "ERROR", 503 to "ERROR")
        for ((statusCode, want) in cases) {
            val span = Telemetry.buildSpan("op", 0, 1, emptyMap(), statusCode)
            @Suppress("UNCHECKED_CAST")
            val status = span["status"] as Map<String, Any?>
            assertEquals(want, status["code"], "status for $statusCode")
        }
    }

    @Test
    fun `emitSpan appends ndjson to the given sink path`(
        @TempDir tempDir: Path,
    ) {
        val sink = tempDir.resolve("spans.ndjson")
        Telemetry.emitSpan(Telemetry.buildSpan("one", 0, 1, emptyMap(), 200), sinkPath = sink.toString())
        Telemetry.emitSpan(Telemetry.buildSpan("two", 2, 3, emptyMap(), 500), sinkPath = sink.toString())

        val lines = Files.readAllLines(sink)
        assertEquals(2, lines.size)
        assertEquals("one", mapper.readTree(lines[0])["name"].asText())
        assertEquals("two", mapper.readTree(lines[1])["name"].asText())
    }

    @Test
    fun `emitSpan writes ndjson to stdout when sinkPath is null`() {
        val out =
            captureStdout {
                Telemetry.emitSpan(Telemetry.buildSpan("stdout-span", 10, 20, emptyMap(), 200), sinkPath = null)
            }
        val node = mapper.readTree(out.trim())
        assertEquals("stdout-span", node["name"].asText())
        assertTrue(out.endsWith("\n"))
    }

    @Test
    fun `emitSpan writes ndjson to stdout when sinkPath is empty`() {
        val out =
            captureStdout {
                Telemetry.emitSpan(Telemetry.buildSpan("stdout-span-2", 10, 20, emptyMap(), 200), sinkPath = "")
            }
        assertEquals("stdout-span-2", mapper.readTree(out.trim())["name"].asText())
    }

    @Test
    fun `emitSpan degrades silently when the sink cannot be opened`(
        @TempDir tempDir: Path,
    ) {
        val badSink = tempDir.resolve("no-such-dir").resolve("spans.ndjson")
        // Must not throw even though the parent directory does not exist.
        Telemetry.emitSpan(Telemetry.buildSpan("dropped", 0, 1, emptyMap(), 200), sinkPath = badSink.toString())
    }

    @Test
    fun `the shipped emitSpan reads the real OTEL_TRACE_FILE env and defaults to stdout`() {
        // OTEL_TRACE_FILE is (virtually certainly) unset in CI/local env -> stdout.
        val out = captureStdout { Telemetry.emitSpan(Telemetry.buildSpan("env-default", 0, 1, emptyMap(), 200)) }
        assertEquals("env-default", mapper.readTree(out.trim())["name"].asText())
    }

    @Test
    fun `recordMetric buckets unknown and non-canonical methods as other`() {
        Telemetry.recordMetric("GET", 200, 5.0)
        Telemetry.recordMetric("BREW", 418, 3.0) // unknown method
        Telemetry.recordMetric("get", 200, 1.0) // case-sensitive: lowercase is not known

        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("http_requests_total{method=\"GET\",status=\"200\"} 1"))
        assertTrue(out.contains("http_requests_total{method=\"other\",status=\"418\"} 1"))
        assertTrue(out.contains("http_requests_total{method=\"other\",status=\"200\"} 1"))
        assertFalse(out.contains("method=\"BREW\""))
        assertFalse(out.contains("method=\"get\""))
    }

    @Test
    fun `recordMetric buckets a null method as other`() {
        Telemetry.recordMetric(null, 200, 1.0)
        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("http_requests_total{method=\"other\",status=\"200\"} 1"))
    }

    @Test
    fun `recordMetric counter aggregates repeated calls`() {
        Telemetry.recordMetric("POST", 201, 2.0)
        Telemetry.recordMetric("POST", 201, 2.0)
        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("http_requests_total{method=\"POST\",status=\"201\"} 2"))
    }

    @Test
    fun `renderMetrics emits HELP and TYPE lines and a trailing newline`() {
        Telemetry.recordMetric("GET", 200, 1.0)
        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("# HELP http_requests_total"))
        assertTrue(out.contains("# TYPE http_requests_total counter"))
        assertTrue(out.contains("# HELP http_request_duration_seconds_total"))
        assertTrue(out.contains("# TYPE http_request_duration_seconds_total counter"))
        assertTrue(out.endsWith("\n"))
    }

    @Test
    fun `renderMetrics on empty state is still valid exposition`() {
        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("# TYPE http_requests_total counter"))
        assertTrue(out.contains("http_request_duration_seconds_total 0"))
    }

    @Test
    fun `renderMetrics duration accumulates`() {
        Telemetry.recordMetric("GET", 200, 500.0) // 0.5s
        Telemetry.recordMetric("GET", 200, 250.0) // 0.25s -> total 0.75s
        val out = Telemetry.renderMetrics()
        assertTrue(out.contains("http_request_duration_seconds_total 0.75"))
    }

    @Test
    fun `renderMetrics emits series in a stable method-then-status order`() {
        Telemetry.recordMetric("POST", 201, 1.0)
        Telemetry.recordMetric("GET", 500, 1.0)
        Telemetry.recordMetric("GET", 200, 1.0)
        val out = Telemetry.renderMetrics()
        val getIdx200 = out.indexOf("method=\"GET\",status=\"200\"")
        val getIdx500 = out.indexOf("method=\"GET\",status=\"500\"")
        val postIdx = out.indexOf("method=\"POST\",status=\"201\"")
        assertTrue(getIdx200 in 0 until getIdx500)
        assertTrue(getIdx500 in 0 until postIdx)
    }

    @Test
    fun `resetMetrics clears request series and zeroes duration`() {
        Telemetry.recordMetric("GET", 200, 5.0)
        Telemetry.resetMetrics()
        val out = Telemetry.renderMetrics()
        assertFalse(out.contains("http_requests_total{"))
        assertTrue(out.contains("http_request_duration_seconds_total 0"))
    }

    @Test
    fun `escapeLabelValue escapes backslash quote and newline`() {
        assertEquals("GET", Telemetry.escapeLabelValue("GET"))
        assertEquals("a\\\"b", Telemetry.escapeLabelValue("a\"b"))
        assertEquals("a\\\\b", Telemetry.escapeLabelValue("a\\b"))
        assertEquals("a\\nb", Telemetry.escapeLabelValue("a\nb"))
    }

    @Test
    fun `escapeLabelValue escapes hostile input`() {
        val hostile = "evil\"\\\n} injected 999"
        val escaped = Telemetry.escapeLabelValue(hostile)
        assertFalse(escaped.replace("\\\"", "").contains("\""))
        assertFalse(escaped.contains("\n"))
        assertEquals("evil\\\"\\\\\\n} injected 999", escaped)
    }

    @Test
    fun `log emits a structured JSON line with ts level and service`() {
        val out = captureStdout { Telemetry.log(mapOf("request_id" to "abc", "status" to 200), serviceName = null) }
        val record = mapper.readTree(out.trim())
        assertEquals("info", record["level"].asText())
        assertEquals("reference-app", record["service"].asText())
        assertEquals("abc", record["request_id"].asText())
        assertEquals(200, record["status"].asInt())
        assertTrue(record.has("ts"))
        // ts must be RFC3339/ISO-8601 parseable.
        Instant.parse(record["ts"].asText())
    }

    @Test
    fun `log falls back to the default service name when serviceName is empty`() {
        val out = captureStdout { Telemetry.log(mapOf("event" to "boot"), serviceName = "") }
        assertEquals("reference-app", mapper.readTree(out.trim())["service"].asText())
    }

    @Test
    fun `log uses the given serviceName when set`() {
        val out = captureStdout { Telemetry.log(mapOf("event" to "boot"), serviceName = "checkout-svc") }
        assertEquals("checkout-svc", mapper.readTree(out.trim())["service"].asText())
    }

    @Test
    fun `the shipped log reads the real SERVICE_NAME env and defaults to reference-app`() {
        // SERVICE_NAME is (virtually certainly) unset in CI/local env -> default.
        val out = captureStdout { Telemetry.log(mapOf("event" to "boot")) }
        assertEquals("reference-app", mapper.readTree(out.trim())["service"].asText())
    }

    @Test
    fun `log ts rejects a garbage value as a sanity check on the format`() {
        // Documents the expectation the positive assertions above rely on: a
        // non-RFC3339 string must fail Instant.parse.
        try {
            Instant.parse("not-a-timestamp")
            error("expected DateTimeParseException")
        } catch (e: DateTimeParseException) {
            // expected
        }
    }
}
